python/arcticdb/version_store/library.py

```diff
@@ -288,6 +302,195 @@ class ReadInfoRequest(NamedTuple):
     symbol: str
     as_of: Optional[AsOf] = None
 
+    def __repr__(self):
+        res = f"ReadInfoRequest(symbol={self.symbol}"
+        res += f", as_of={self.as_of}" if self.as_of is not None else ""
+        res += ")"
+        return res
+
+
+class LazyDataFrame(QueryBuilder):
+    """
+    Lazy dataframe implementation, allowing chains of queries to be added before the read is actually executed.
+    Returned by `Library.read`, `Library.head`, and `Library.tail` calls when `lazy=True`.
+
+    See Also
+    --------
+    QueryBuilder for supported querying operations.
+
+    Examples
+    --------
+
+    # Specify that we want version 0 of "test" symbol, and to only return the "new_column" column in the output
+    >>> lazy_df = lib.read("test", as_of=0, columns=["new_column"], lazy=True)
+    # Perform a filtering operation
+    >>> lazy_df = lazy_df[lazy_df["col1"].isin(0, 3, 6, 9)]
+    # Create a new column through a projection operation
+    >>> lazy_df["new_col"] = lazy_df["col1"] + lazy_df["col2"]
+    # Actual read and processing happens here
+    >>> df = lazy_df.collect().data
+    """
+    def __init__(
+            self,
+            lib: "Library",
+            read_request: ReadRequest,
+    ):
+        if read_request.query_builder is None:
+            super().__init__()
+        else:
+            self.clauses = read_request.query_builder.clauses
+            self._python_clauses = read_request.query_builder._python_clauses
+            self._optimisation = read_request.query_builder._optimisation
+        self.lib = lib
+        self.read_request = read_request._replace(query_builder=None)
+
+    def to_read_request(self) -> ReadRequest:
+        """
+        Convert this object into a ReadRequest, including any queries applied to this object since the read call.
+
+        Returns
+        -------
+        ReadRequest
+            Object with all the parameters necessary to completely specify the data to be read.
+        """
+        q = QueryBuilder().prepend(self)
+        q._optimisation = self._optimisation
+        return ReadRequest(
+            symbol=self.read_request.symbol,
+            as_of=self.read_request.as_of,
+            date_range=self.read_request.date_range,
+            row_range=self.read_request.row_range,
+            columns=self.read_request.columns,
+            query_builder=q,
+        )
+
+    def collect(self) -> VersionedItem:
+        """
+        Read the data and execute any queries applied to this object since the read call.
+
+        Returns
+        -------
+        VersionedItem
+            Object that contains a .data and .metadata element.
+        """
+        return self.lib.read(**self.to_read_request()._asdict())
+
+    def __str__(self) -> str:
+        query_builder_repr = super().__str__()
+        return "LazyDataFrame(" + self.read_request.__repr__() + (" | " if len(query_builder_repr) else "") + query_builder_repr + ")"
```

Comment:
Why `if len(query_builder_repr)` not `if query_builder_repr` ?

Reply:
Done